ProjFS.Mac/PrjFSKext/FileSystemMountPoints.cpp

```diff
@@ -0,0 +1,131 @@
+#include "FileSystemMountPointsPrivate.hpp"
+#include "FileSystemMountPoints.hpp"
+#include "kernel-header-wrappers/mount.h"
+#include "Locks.hpp"
+#include "KextLog.hpp"
+#include <kern/assert.h>
+
+static Mutex s_mountPointsMutex;
+KEXT_STATIC MountPoint s_usedMountPoints[MaxUsedMountPoints];
+
+bool MountPoint_Init()
+{
+    s_mountPointsMutex = Mutex_Alloc();
+    if (!Mutex_IsValid(s_mountPointsMutex))
+    {
+        return false;
+    }
+    
+    return true;
+}
+
+void MountPoint_Cleanup()
+{
+    if (Mutex_IsValid(s_mountPointsMutex))
+    {
+        Mutex_FreeMemory(&s_mountPointsMutex);
+    }
+    
+    for (size_t i = 0; i < MaxUsedMountPoints; ++i)
+    {
+        // by the time this function is called, any provider user clients
+        // should have been terminated & deregistered
+        assert(s_usedMountPoints[i].mountPoint == nullptr);
+        assert(s_usedMountPoints[i].authCacheDisableCount == 0);
+    }
+}
+
+KEXT_STATIC ssize_t FindMountPoint(mount_t mountPoint)
```

Comment:
Obviously we can only call this under lock.  Is there any way to enforce that?

Reply:
Not conventionally, but I've clearly forgotten to apply the naming convention of suffixing with `Locked` which I myself introduced to other parts of the project. Oops.

Re "not conventionally": [Rust's mutex implementation](https://doc.rust-lang.org/1.0.0/book/concurrency.html#safe-shared-mutable-state) is interesting in this regard - the mutex is a generic, templated on the type of the object it is protecting. (or an aggregate of the multiple objects it is protecting) Acquiring the mutex returns a reference to the protected object for the duration of the held lock. Rust's type system (a limited version of linear typing) is powerful enough to prevent you from holding onto that reference outside of the locked block.

You could replicate that same programming pattern in C++ by turning our `Mutex` struct into a template and making lock acquisition scoped:

```
struct UsedMountPoints
{
    MountPoint mountPoints[MaxUsedMountPoints];
};

static Mutex<UsedMountPoints> s_mountPointsMutex;

// …

ssize_t FindMountPoint(MutexGuard<UsedMountPoints>& lockedMountPoints, mount_t mountPoint)
{
  for (size_t i = 0; i < MaxUsedMountPoints; ++i)
  {
    if (lockedMountPoints->usedMountPoints[i].mountPoint == mountPoint)
    {
      return i;
    }
  }

  return -1;
}

void MountPoint_DisableAuthCache(mount_t mountPoint)
{
  {
    MutexGuard<UsedMountPoints> lockedMountPoints(s_mountPointsMutex);
    ssize_t mountPointIndex = FindMountPoint(lockedMountPoints, mountPoint);
  }
} 
```

`FindMountPoint` can now only be called by a lock holder as it expects a `MutexGuard` reference.

In the expression `lockedMountPoints->usedMountPoints[i]` you can see that I'm implementing `operator->` on the `MutexGuard` to return `UsedMountPoints*` which can safely be mutated because the guard is holding the lock.

Or you could perhaps use some slightly different notation.

You could still break out of this by explicitly calling `operator->` and saving that pointer beyond where the guard goes out of scope, but this should in theory be more a much more obvious transgression than accidentally accessing an informally mutex-protected variable outside of a lock.

Whether or not we want to apply this sort of pattern to the kext code is another question of course!

Reply:
The latest push suffixes the function names with `_Locked` as appropriate.